ruma-client-api/src/r0/session/get_login_types.rs

```diff
@@ -40,51 +48,345 @@ impl Response {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// The payload for SSO login.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[serde(tag = "type", rename = "m.login.sso")]
+pub struct SsoLoginType {
+    /// The identity provider choices.
+    ///
+    /// This uses the unstable prefix in
+    /// [MSC2858](https://github.com/matrix-org/matrix-doc/pull/2858).
+    #[cfg(feature = "unstable-pre-spec")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+    #[serde(rename = "org.matrix.msc2858.identity_providers")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub identity_providers: Option<Vec<IdentityProvider>>,
+}
+
+impl SsoLoginType {
+    /// Creates a new `PasswordLoginType`.
+    pub fn new() -> Self {
+        Self {
+            #[cfg(feature = "unstable-pre-spec")]
+            identity_providers: None,
+        }
+    }
+}
+
+impl Default for SsoLoginType {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// An SSO login identity provider.
+#[cfg(feature = "unstable-pre-spec")]
+#[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct IdentityProvider {
+    /// The ID of the provider.
+    id: String,
+
+    /// The display name of the provider.
+    name: String,
+
+    /// The icon for the provider.
+    icon: Option<MxcUri>,
+
+    /// The brand identifier for the provider.
+    brand: Option<IdentityProviderBrand>,
+}
+
+#[cfg(feature = "unstable-pre-spec")]
+impl IdentityProvider {
+    /// Creates an `IdentityProvider` with the given `id` and `name`.
+    pub fn new(id: &str, name: &str) -> Self {
+        Self { id: id.into(), name: name.into(), icon: None, brand: None }
+    }
+}
+
+/// An SSO login identity provider brand identifier.
+///
+/// This uses the unstable prefix in
+/// [MSC2858](https://github.com/matrix-org/matrix-doc/pull/2858).
+#[cfg(feature = "unstable-pre-spec")]
+#[cfg_attr(docsrs, doc(cfg(feature = "unstable-pre-spec")))]
+#[derive(Clone, Debug, PartialEq, Eq, StringEnum)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum IdentityProviderBrand {
+    /// The [Apple] brand.
+    ///
+    /// [Apple]: https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/
+    #[ruma_enum(rename = "org.matrix.apple")]
+    Apple,
+
+    /// The [Facebook](https://developers.facebook.com/docs/facebook-login/web/login-button/) brand.
+    #[ruma_enum(rename = "org.matrix.facebook")]
+    Facebook,
+
+    /// The [GitHub](https://github.com/logos) brand.
+    #[ruma_enum(rename = "org.matrix.github")]
+    Github,
+
+    /// The [GitLab](https://about.gitlab.com/press/press-kit/) brand.
+    #[ruma_enum(rename = "org.matrix.gitlab")]
+    Gitlab,
```

Comment:
I think we should use the same capitalization used by the services themselves since it doesn't really clash with anything and looks nicer.